api/trace/current_test.go

```diff
@@ -0,0 +1,110 @@
+package trace_test
+
+import (
+	"context"
+	"testing"
+
+	"google.golang.org/grpc/codes"
+
+	"go.opentelemetry.io/api/core"
+	"go.opentelemetry.io/api/event"
+	"go.opentelemetry.io/api/tag"
+	"go.opentelemetry.io/api/trace"
+)
+
+func TestSetCurrentSpanOverridesPreviouslySetSpan(t *testing.T) {
+	originalSpan := trace.NoopSpan{}
+	expectedSpan := mockSpan{}
+
+	ctx := context.Background()
+
+	ctx = trace.SetCurrentSpan(ctx, originalSpan)
+	ctx = trace.SetCurrentSpan(ctx, expectedSpan)
+
+	if span := trace.CurrentSpan(ctx); span != expectedSpan {
+		t.Errorf("Want: %v, but have: %v", expectedSpan, span)
+	}
+}
+
+func TestCurrentSpan(t *testing.T) {
+	for _, testcase := range []struct {
+		name string
+		ctx  context.Context
+		want trace.Span
+	}{
+		{
+			name: "CurrentSpan() returns a NoopSpan{} from an empty context",
+			ctx:  context.Background(),
+			want: trace.NoopSpan{},
+		},
+		{
+			name: "CurrentSpan() returns current span if set",
+			ctx:  trace.SetCurrentSpan(context.Background(), mockSpan{}),
+			want: mockSpan{},
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			// proto: CurrentSpan(ctx context.Context) trace.Span
+			have := trace.CurrentSpan(testcase.ctx)
+			if have != testcase.want {
+				t.Errorf("Want: %v, but have: %v", testcase.want, have)
+			}
+		})
+	}
+}
+
+// a duplicate of trace.NoopSpan for testing
+type mockSpan struct{}
```

Comment:
I support exporting `trace.noopSpan` in order to avoid this block of code.

Reply:
So.. I think my comment is misleading. Because several methods return NoopSpan as a fallback-behavior, the tests are still going to need a mocked-up `Span` type to differentiate successful invocations from fallback behavior.  See for example: https://github.com/open-telemetry/opentelemetry-go/blob/b21a60af81127a83eee1e757f69f43c6bd8c3cea/api/trace/current_test.go#L17